src/posix/net.cpp

```diff
@@ -2018,20 +2089,24 @@ bool CurlHttpIO::multidoio(CURLM *curlmhandle)
                         ipv6deactivationtime = Waiter::ds;
 
                         // for IPv6 errors, try IPv4 before sending an error to the engine
-                        if((dnsEntry.ipv4.size() && Waiter::ds - dnsEntry.ipv4timestamp < DNS_CACHE_TIMEOUT_DS) || httpctx->ares_pending)
+                        if ((dnsEntry.ipv4.size()
+                            && (!DNS_CACHE_EXPIRES || (Waiter::ds - dnsEntry.ipv4timestamp) < DNS_CACHE_TIMEOUT_DS))
+                                || httpctx->ares_pending)
```

Comment:
Interesting case. Until this pull request, it shouldn't be possible because the current request is IPv6 here so, if there was a DNS request, the IPv6 should have finished (otherwise the IP was cached and therefore there wasn't any IPv6 DNS request) or the request couldn't have been sent yet because the DNS request (IPv6) should have finished before.
However, with this pull request, a DNS can be sent even if the IP was cached (to check if it's invalid) so it might happen. I'm going to check it.

Reply:
I think that it's very unlikely to hit that case because if that happens the IPv4 should be cached but I'm going to add an additional check to only apply the optimization of the IP was not cached (like it was working before). That way, the code should be totally safe and, in that unlikely case, the connection error would be just forwarded to the upper layer and managed as usual. Finally, this is just an optimization for a very specific case, so I think that it's OK.